examples_apps/run_tutorial_machine_learning.py

```diff
@@ -0,0 +1,319 @@
+#     Out:
+#
+#     .. code-block:: none
+#
+#         ParameterError: {alpha}: unbound parameter with no default value.
+#
+# To bind a numerical value to the free parameters, we pass a mapping from parameter
+# name to value using the ``args`` keyword argument when running the engine:
+
+mapping = {
+    "alpha": tf.Variable(0.5),
+    "theta_bs": tf.constant(0.4),
+    "phi_bs": tf.constant(0.0)
+}
+
+result = eng.run(prog, args=mapping)
+
+
+##############################################################################
+# This code will execute without error, and both the output results and the
+# output samples will contain numeric values based on the given value for the angle ``phi``:
+
+print(result.samples)
+
+##############################################################################
+# We can select measurement results at other angles by supplying different values for
+# ``phi`` in ``mapping``. We can also return and perform processing on the underlying state:
+
+state = result.state
+print("Density matrix element [0,0,1,2]:", state.dm()[0, 0, 1, 2])
+
+##############################################################################
+# Computing gradients
+# -------------------
+#
+# When used within a gradient tape context, any output generated from the state
+# class can be differentiated natively using ``GradientTape().gradient()``.
+# For example, lets displace a one-mode program, and then compute the gradient
+# of the mean photon number:
+
+eng.reset()
+prog = sf.Program(1)
+
+alpha = prog.params("alpha")
+
+with prog.context as q:
+    Dgate(alpha) | q
+
+# Assign our TensorFlow variables, so that we can
+# refer to them later when differentiating/training.
+a = tf.Variable(0.43)
+
+with tf.GradientTape() as tape:
+    # Here, we pass map our quantum free parameter `alpha`
+    # to our TensorFlow variable `a`.
+    result = eng.run(prog, args={"alpha": a})
+    state = result.state
+
+    # Note that all processing, even state-based post-processing,
+    # must be done within the gradient tape!
+    mean, var = state.mean_photon(0)
+
+# test the gradient of the variance is correct
+grad = tape.gradient(mean, [a])
+print("Gradient:", grad)
+
+##############################################################################
+# The mean photon number of a displaced state :math:`|\alpha\rangle` is given by
+# :math:`\langle \hat{n} \rangle(\alpha) = |\alpha|^2`, and thus, for real-valued
+# :math:`\alpha`, the gradient is given by :math:`\langle \hat{n}\rangle'(\alpha) = 2\alpha`:
+
+print("Exact gradient:", 2*a)
+print("Exact and TensorFlow gradient agree:", np.allclose(grad, 2*a))
+
+##############################################################################
+# Processing data
+# ---------------
+#
+# The parameters for Blackbird states, gates, and measurements may be more complex
+# than just raw data or machine learning weights. These can themselves be the outputs
+# from some learnable function like a neural network [#]_.

```

Comment:
Because it was written using TensorFlow 1.0, and I'm not sure it is relevant any more (but thought it was nice to keep the intro text). In TensorFlow 1.0, creating the neural network was more complicated, as you were constructing a symbolic and unevaluated graph using `tf.placeholders`. Now, creating the neural network is no more complicated than simply matrix multiplication, and feeding the output of that to the Dgate.